cpp/src/arrow/csv/writer.cc

```diff
@@ -328,6 +351,26 @@ class QuotedColumnPopulator : public ColumnPopulator {
   }
 
  private:
+  // Returns true if there's no quote in the string array
+  // similar to std::find, but with much better performance
+  static bool NoQuoteInArray(const StringArray& array) {
+    const uint8_t* data = array.raw_data() + array.value_offset(0);
+    const int64_t buffer_size = array.total_values_length();
+    const uint8_t* const data_end = data + buffer_size;
+
+    for (int64_t i = 0; i < buffer_size / 16; ++i) {
+      bool r = false;
+      for (int i = 0; i < 16; ++i) {
+        r |= data[i] == '"';
+      }
```

Comment:
How does this compare to simply using `memchr`? At least on glibc it should be heavily optimized. I have no idea on macOS or Windows.

Reply:
Does using CountQuotes here hurt performance?

Reply:
`memchr` is great per my test. Will use it. Don't know why `std::find` doesn't vectorize the code.
`CountQuotes` hurts performance a bit (about 30% drop for StringNoQuote/0). Guess it does more work than necessary.

Reply:
`std::string_view::find` and `std::char_traits<char>::find` calls `memchr` internaly. But `std::find`doesn't.
https://godbolt.org/z/asGx1f35s